spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitMessageChannelBinder.java

```diff
@@ -194,21 +172,64 @@ public RabbitConsumerProperties getExtendedConsumerProperties(String channelName
 		return this.extendedBindingProperties.getExtendedConsumerProperties(channelName);
 	}
 
-
 	@Override
 	public RabbitProducerProperties getExtendedProducerProperties(String channelName) {
 		return this.extendedBindingProperties.getExtendedProducerProperties(channelName);
 	}
 
 	@Override
-	protected MessageProducer createConsumerEndpoint(String name, String group, Queue destination,
-			ExtendedConsumerProperties<RabbitConsumerProperties> properties) {
+	protected MessageHandler createProducerMessageHandler(final ProducerDestination producerDestination,
+														ExtendedProducerProperties<RabbitProducerProperties> producerProperties)
+			throws Exception {
+		String prefix = producerProperties.getExtension().getPrefix();
+		String exchangeName = producerDestination.getName();
+		String destination = StringUtils.isEmpty(prefix) ? exchangeName : exchangeName.substring(prefix.length());
+		final AmqpOutboundEndpoint endpoint = new AmqpOutboundEndpoint(buildRabbitTemplate(producerProperties.getExtension()));
+		endpoint.setExchangeName(producerDestination.getName());
+		RabbitProducerProperties extendedProperties = producerProperties.getExtension();
+		String routingKeyExpression = extendedProperties.getRoutingKeyExpression();
+		if (!producerProperties.isPartitioned()) {
+			if (routingKeyExpression == null) {
+				endpoint.setRoutingKey(destination);
+			}
+			else {
+				endpoint.setRoutingKeyExpressionString(routingKeyExpression);
+			}
+		}
+		else {
+			if (routingKeyExpression == null) {
+				endpoint.setRoutingKeyExpressionString(buildPartitionRoutingExpression(destination));
+			}
+			else {
+				endpoint.setRoutingKeyExpressionString(buildPartitionRoutingExpression(routingKeyExpression));
```

Comment:
Looks like omission:
The `buildPartitionRoutingExpression()` has code like:
```
protected String buildPartitionRoutingExpression(String expressionRoot) {
	return "'" + expressionRoot + "-' + headers['" + BinderHeaders.PARTITION_HEADER + "']";
}
```
So, we concatenate provided string with some header.
Not sure that full provided SpEL should be treat as `literal`.
I guess you meant here something pretty straightforward:
```
endpoint.setRoutingKeyExpressionString(routingKeyExpression);
```

No particular test or am I missing something in the logic?
Thanks